File: lib/reputable_chat/cryptography/seed.rb
# frozen_string_literal: true

require "digest"

module ReputableChat
  module Cryptography
    # BIP39 wordlist encoding and checksum.
    #
    # The server never sees a seed -- this exists as the reference
    # implementation that public/js/seed.js must match byte for byte, and as
    # the source of the shared test vectors in spec/seed_spec.rb.
    #
    # Sizing (see docs/project/identity.md): 2048 words is 11 bits each, and 8
    # bits go to the checksum. Eight words is therefore 80 bits of entropy.
    # Because there is no username to salt with, an attacker grinds candidate
    # seeds against every registered public key at once, so the cost of
    # breaking SOME account is 2**80 / users rather than 2**80. At 10M users
    # that is ~457 years against a memory-hard KDF. Seven words would be ~81
    # days. The KDF is not optional at any length.
    module Seed
      WORDLIST_PATH = File.expand_path("../../../config/bip39-english.txt", __dir__)
      BITS_PER_WORD = 11
      CHECKSUM_BITS = 8
      MIN_WORDS     = 8

      class InvalidSeed < StandardError; end

      module_function

      def wordlist
        @wordlist ||= File.readlines(WORDLIST_PATH, chomp: true).map(&:strip).freeze
      end

      def word_index
        @word_index ||= wordlist.each_with_index.to_h.freeze
      end

      # Normalized form: lowercase, single-spaced. Both the checksum and the
      # KDF run over this, so whitespace and case can never change an identity.
      def normalize(phrase)
        phrase.to_s.downcase.strip.split(/\s+/).join(" ")
      end

      def words(phrase) = normalize(phrase).split(" ")

      def valid?(phrase, min_words: MIN_WORDS)
        validate!(phrase, min_words: min_words)
        true
      rescue InvalidSeed
        false
      end

      # Raises with a reason the UI can show. Distinguishing "not a word" from
      # "checksum failed" matters: the first is a typo the user can fix, the
      # second means they have transposed or mis-remembered something.
      def validate!(phrase, min_words: MIN_WORDS)
        list = words(phrase)

        raise InvalidSeed, "a seed needs at least #{min_words} words" if list.size < min_words

        unknown = list.reject { |w| word_index.key?(w) }
        raise InvalidSeed, "not in the wordlist: #{unknown.join(', ')}" if unknown.any?

        raise InvalidSeed, "checksum failed - check for a mistyped or swapped word" unless checksum_ok?(list)

        true
      end

      def checksum_ok?(list)
        bits = list.map { |w| word_index.fetch(w).to_s(2).rjust(BITS_PER_WORD, "0") }.join
        entropy_bits = bits[0...-CHECKSUM_BITS]
        provided     = bits[-CHECKSUM_BITS..]

        provided == checksum_for(entropy_bits, list.size)
      end

      # The word count is hashed alongside the entropy so that seeds of
      # different lengths cannot collide once the entropy is byte-padded.
      def checksum_for(entropy_bits, word_count)
        padded = entropy_bits.ljust((entropy_bits.length / 8.0).ceil * 8, "0")
        bytes  = [padded].pack("B*")
        digest = Digest::SHA256.digest([word_count].pack("C") + bytes)

        digest.unpack1("B*")[0, CHECKSUM_BITS]
      end

      # Builds a valid phrase from raw entropy bits. Used by the tests and by
      # the client when generating a new seed.
      def encode(entropy_bits, word_count)
        bits = entropy_bits + checksum_for(entropy_bits, word_count)
        bits.chars.each_slice(BITS_PER_WORD).map { |slice| wordlist.fetch(slice.join.to_i(2)) }.join(" ")
      end

      def entropy_bits_for(word_count) = (word_count * BITS_PER_WORD) - CHECKSUM_BITS
    end
  end
end

File: public/js/seed.js
// BIP39 wordlist encoding and checksum. Mirrors
// lib/reputable_chat/cryptography/seed.rb.

const BITS_PER_WORD = 11;
const CHECKSUM_BITS = 8;
export const MIN_WORDS = 8;

let WORDLIST = null;
let INDEX = null;

export async function loadWordlist(url = "/wordlist.txt") {
  if (WORDLIST) return WORDLIST;

  const response = await fetch(url);
  if (!response.ok) throw new Error("could not load the wordlist");

  WORDLIST = (await response.text()).split("\n").map((w) => w.trim()).filter(Boolean);
  if (WORDLIST.length !== 2048) throw new Error(`wordlist should hold 2048 words, got ${WORDLIST.length}`);

  INDEX = new Map(WORDLIST.map((word, i) => [word, i]));
  return WORDLIST;
}

// Both the checksum and the key derivation run over this form, so whitespace
// and capitalisation can never change an identity.
export function normalize(phrase) {
  return String(phrase || "").toLowerCase().trim().split(/\s+/).filter(Boolean).join(" ");
}

export function words(phrase) {
  const cleaned = normalize(phrase);
  return cleaned ? cleaned.split(" ") : [];
}

export function entropyBitsFor(wordCount) {
  return wordCount * BITS_PER_WORD - CHECKSUM_BITS;
}

// Returns null when the phrase is good, or a reason the UI can show.
// Distinguishing "not a word" from "checksum failed" matters: the first is a
// typo the user can see and fix, the second means a word is transposed or
// misremembered and they need to look harder.
export async function validate(phrase, minWords = MIN_WORDS) {
  await loadWordlist();
  const list = words(phrase);

  if (list.length < minWords) return `A seed needs at least ${minWords} words.`;

  const unknown = list.filter((w) => !INDEX.has(w));
  if (unknown.length) return `Not in the wordlist: ${unknown.join(", ")}`;

  if (!(await checksumOk(list))) return "Checksum failed — check for a mistyped or swapped word.";

  return null;
}

export async function isValid(phrase, minWords = MIN_WORDS) {
  return (await validate(phrase, minWords)) === null;
}

async function checksumOk(list) {
  const bits = list.map((w) => INDEX.get(w).toString(2).padStart(BITS_PER_WORD, "0")).join("");
  const entropy = bits.slice(0, -CHECKSUM_BITS);
  const provided = bits.slice(-CHECKSUM_BITS);

  return provided === (await checksumFor(entropy, list.length));
}

// The word count is hashed alongside the entropy so seeds of different lengths
// cannot collide once the entropy is byte-padded.
async function checksumFor(entropyBits, wordCount) {
  const padded = entropyBits.padEnd(Math.ceil(entropyBits.length / 8) * 8, "0");
  const entropyBytes = new Uint8Array(padded.length / 8);
  for (let i = 0; i < entropyBytes.length; i++) {
    entropyBytes[i] = parseInt(padded.slice(i * 8, i * 8 + 8), 2);
  }

  const input = new Uint8Array(1 + entropyBytes.length);
  input[0] = wordCount;
  input.set(entropyBytes, 1);

  const digest = new Uint8Array(await crypto.subtle.digest("SHA-256", input));
  return digest[0].toString(2).padStart(8, "0").slice(0, CHECKSUM_BITS);
}

export async function generate(wordCount = MIN_WORDS) {
  await loadWordlist();

  const bitCount = entropyBitsFor(wordCount);
  const random = new Uint8Array(Math.ceil(bitCount / 8));
  crypto.getRandomValues(random);

  const entropy = Array.from(random)
    .map((b) => b.toString(2).padStart(8, "0"))
    .join("")
    .slice(0, bitCount);

  const bits = entropy + (await checksumFor(entropy, wordCount));
  const out = [];
  for (let i = 0; i < bits.length; i += BITS_PER_WORD) {
    out.push(WORDLIST[parseInt(bits.slice(i, i + BITS_PER_WORD), 2)]);
  }
  return out.join(" ");
}
